Adding order (#6)

Wudext opened the pull request.

parfenovma committed:
tests initial with base conftest

parfenovma committed:
button 200 OK API tested

Wudext committed:
Adding order (#6)

Wudext committed:
Update 4c4c9c7f8e24_init.py

dyakovri commented on services_backend/routes/category.py:
```diff
@@ -10,14 +12,18 @@
 @category.get("/", response_model=list[CategoryGet])
 def get_categories(offset: int = 0, limit: int = 100):
-    return db.session.query(Category).offset(offset).limit(limit).all()
+    return sorted(db.session.query(Category).offset(offset).limit(limit).all(), key=operator.attrgetter("order"))
```
У db.session.query(...) можно дописать `.order_by(Category.order)`, будет работать быстрее и лучше работает с лимитами

dyakovri commented on services_backend/routes/category.py:
```diff
@@ -48,9 +58,19 @@ def update_category(category_inp: CategoryUpdate, category_id: int):
     if not any(category_inp.dict().values()):
         raise HTTPException(status_code=400, detail="Empty schema")
+
+    if category.one_or_none().order > category_inp.order:
+        for categ in db.session.query(Category).filter(Category.order < category.one_or_none().order):
+            categ = db.session.query(Category).filter(Category.id == categ.id)
+            categ.update({"order": Category.order + 1})
+
+    elif category.one_or_none().order < category_inp.order:
+        for categ in db.session.query(Category).filter(Category.order > category.one_or_none().order):
+            categ = db.session.query(Category).filter(Category.id == categ.id)
+            categ.update({"order": Category.order - 1})
```
Посмотри по update доку
https://docs.sqlalchemy.org/en/20/core/dml.html#sqlalchemy.sql.expression.update

Можно сделать 
```python
db.session.update(Category)  \
    .filter(Category.order < current_category.order) \
    .values(order = Category.order + 1)
```

Wudext committed:
Update category.py

Wudext committed:
Update category.py

Wudext committed:
Update button.py

Wudext committed:
Update category.py

Wudext committed:
Update button.py

Wudext committed:
Upgrading to SQLAlchemy 2.0

Wudext committed:
Fixing Type-Hints

parfenovma committed:
order type fix + migration for SA2 + fixtures for SA2
dbsession.add(), dbsession.query()....one_or_none(), etc. is legacy code in SA2
